dbms/src/Common/HashTable/StringHashTable.h

```diff
@@ -0,0 +1,481 @@
+{
+    // Smooth growing for string maps
+    void increaseSize() { this->size_degree += 1; }
+};
+
+template <typename SubMaps>
+class StringHashTable : private boost::noncopyable
+{
+protected:
+    static constexpr size_t NUM_MAPS = 5;
+    // Map for storing empty string
+    using T0 = typename SubMaps::T0;
+
+    // Short strings are stored as numbers
+    using T1 = typename SubMaps::T1;
+    using T2 = typename SubMaps::T2;
+    using T3 = typename SubMaps::T3;
+
+    // Long strings are stored as StringRef along with saved hash
+    using Ts = typename SubMaps::Ts;
+    using Self = StringHashTable;
+
+    template <typename, typename, size_t>
+    friend class TwoLevelStringHashTable;
+
+    T0 m0;
+    T1 m1;
+    T2 m2;
+    T3 m3;
+    Ts ms;
+
+public:
+    using Key = StringRef;
+    using key_type = Key;
+    using value_type = typename Ts::value_type;
+
+    StringHashTable() {}
+
+    StringHashTable(size_t reserve_for_num_elements)
+        : m1{reserve_for_num_elements / 4}
+        , m2{reserve_for_num_elements / 4}
+        , m3{reserve_for_num_elements / 4}
+        , ms{reserve_for_num_elements / 4}
+    {
+    }
+
+    StringHashTable(StringHashTable && rhs) { *this = std::move(rhs); }
+    ~StringHashTable() {}
+
+public:
+    // Dispatch is written in a way that maximizes the performance:
+    // 1. Always memcpy 8 times bytes
+    // 2. Use switch case extension to generate fast dispatching table
+    // 3. Combine hash computation along with key loading
+    // 4. Funcs are named callables that can be force_inlined
+    // NOTE: It relies on Little Endianness and SSE4.2
+    template <typename Func>
+    decltype(auto) ALWAYS_INLINE dispatch(Key & x, Func && func)
+    {
+        static constexpr StringKey0 key0{};
+        size_t sz = x.size;
+        const char * p = x.data;
+        // pending bits that needs to be shifted out
+        char s = (-sz & 7) * 8;
+        size_t res = -1ULL;
+        UInt64 n[3];
+        switch (sz)
+        {
+            case 0:
+                return func(m0, key0, 0);
+            CASE_1_8 : {
+                // first half page
+                if ((reinterpret_cast<uintptr_t>(p) & 2048) == 0)
+                {
+                    memcpy(&n[0], p, 8);
+                    n[0] &= -1ul >> s;
+                }
+                else
+                {
+                    const char * lp = x.data + x.size - 8;
+                    memcpy(&n[0], lp, 8);
+                    n[0] >>= s;
+                }
+                res = _mm_crc32_u64(res, n[0]);
+                return func(m1, n[0], res);
+            }
+            CASE_9_16 : {
+                memcpy(&n[0], p, 8);
+                res = _mm_crc32_u64(res, n[0]);
+                const char * lp = x.data + x.size - 8;
+                memcpy(&n[1], lp, 8);
+                n[1] >>= s;
+                res = _mm_crc32_u64(res, n[1]);
+                return func(m2, *reinterpret_cast<StringKey16 *>(n), res);
```

Comment:
This also leads to a PVS-Studio warning that size of n is not a multiple of size of StringKey16, but I don't know how to get rid of it. I tried &n[0] and the warning is still there.

Reply:
Hmm, I don't have PVS-Studio available. I'll make some attempt first. Could you try the updated commit? I've applied for a license of PVS-Studio and will check it locally after available.

Reply:
We actually have it in the CI, so it is run for all pull requests. If you click "Show all check" at the bottom of the PR page, you can see the PVS-Studio check results, too. Here it is for your latest commit: https://clickhouse-test-reports.s3.yandex.net/5417/87f7fd2290d39d74bebb0cac807978afe3001fd7/pvs_studio_report/pvs-studio-html-report/index.html

The commit that I tried is in this branch: https://github.com/yandex/ClickHouse/commits/aku/tmp There's a yellow circle to the right of the last commit, if you click it there is also a link to PVS-Studio report.

Reply:
Ah, thanks. My network to Github is extremely unreliable now thus I cannot push my last commit. I'll try again tommorrow.

Reply:
Now that warning is gone.